clarity/src/vm/analysis/type_checker/tests/contracts.rs

```diff
@@ -630,3 +664,1864 @@ fn test_expects() {
         _ => false,
     });
 }
+
+#[test]
+fn test_traits() {
+    {
+        let trait_to_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-1 (contract <trait-1>))
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+
+        let trait_to_compatible_trait = "(define-trait trait-1 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (echo (uint) (response uint uint))
+        ))
+        (define-public (wrapped-echo (contract <trait-1>))
+            (internal-echo contract))
+        (define-public (internal-echo (contract <trait-2>))
+            (ok true))";
+
+        let bad_principal_to_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-1 (contract principal))
+            (internal-get-1 contract))
+        (define-public (internal-get-1 (contract <trait-1>))
+            (contract-call? contract get-1 u1))";
+
+        let bad_other_trait = "(define-trait trait-1 (
+            (get-1 (uint) (response uint uint))
+        ))
+        (define-trait trait-2 (
+            (get-2 (uint) (response uint uint))
+        ))
+        (define-public (wrapped-get-2 (contract <trait-1>))
+            (internal-get-2 contract))
+        (define-public (internal-get-2 (contract <trait-2>))
+            (contract-call? contract get-2 u1))";
+
```

Comment:
All of these test cases are pretty hard to follow.  Can you add a comment on each one to explain what code path is being tested (or why the code is bad)?

Also, I don't see any tests to verify that you can't nest traits too deeply.

Reply:
Refactored and added comments in f3be64c28

Reply:
I will add a new test for the deep nesting.

Reply:
I don't think there should be anything unique about deep nesting with a trait versus with any other type. I think that a deep nesting trait test wouldn't make it past the parser to the type-checker, right?